core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,225 @@
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+
+/**
+ * In an audit workflow, new data is written to an orphan snapshot that is not committed as the table's
+ * current state until it is audited. After auditing a change, it may need to be applied or cherry-picked
+ * on top of the latest snapshot instead of the one that was current when the audited changes were created.
+ *
+ * This class adds support for cherry-picking the changes from an orphan snapshot by applying them to
+ * the current snapshot. The output of the operation is a new snapshot with the changes from cherry-picked
+ * snapshot.
+ *
+ * Cherry-picking should apply the exact set of changes that were done in the original commit.
+ *  - All added files should be added to the new version.
+ *  - Todo: If files were deleted, then those files must still exist in the data set.
+ *  - Does not support Overwrite operations currently. Overwrites are considered as conflicts.
+ *
+ */
+class CherryPickFromSnapshot extends MergingSnapshotProducer<AppendFiles> implements CherryPick {
+  private final TableOperations ops;
+  private TableMetadata base = null;
+  private Long cherryPickSnapshotId = null;
+  private final AtomicInteger manifestCount = new AtomicInteger(0);
+
+  CherryPickFromSnapshot(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.refresh();
+  }
+
+  @Override
+  protected AppendFiles self() {
+    throw new UnsupportedOperationException("Not supported");
+  }
+
+  /**
+   * We only cherry pick for appends right now
+   * @return
+   */
+  @Override
+  protected String operation() {
+    return DataOperations.APPEND;
+  }
+
+  @Override
+  public CherryPickFromSnapshot fromSnapshotId(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    this.cherryPickSnapshotId = snapshotId;
+    return this;
+  }
+
+  /**
+   * Apply the pending changes and return the uncommitted changes for validation.
+   * <p>
+   * This does not result in a permanent update.
+   *
+   * @return the uncommitted changes that would be committed by calling {@link #commit()}
+   * @throws ValidationException      If the pending changes cannot be applied to the current metadata
+   * @throws IllegalArgumentException If the pending changes are conflicting or invalid
+   */
+  @Override
+  public Snapshot apply() {
+    ValidationException.check(cherryPickSnapshotId != null,
+        "Cannot cherry pick unknown version: call fromSnapshotId");
+
+    Snapshot cherryPickSnapshot = base.snapshot(cherryPickSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    ValidationException.check(wapId != null,
+        "Snapshot provided for cherrypick didn't have wap.id set on it. " +
+            "Only snapshots that were staged can be cherrypicked from.");
```

Comment:
Staged snapshots don't necessarily need to be created by WAP. Why require this?

Reply:
We can take this out, things is we didn't have the check for duplicate published snapshot earlier, so as a consequence here we are trying to ensure that the snapshot being cherrypicked was indeed staged before. i'l take this out. 

But on a related note, in what scenario would a staged snapshot not have a wap id? I thought the `Writer` sets the wap.id when staging the snapshot. Maybe you'r referring to future possible usecase of staging snapshots?

Reply:
This doesn't necessarily need to be used with staged snapshots. It is also useful for picking changes that were reverted by rollbacks. Say I have a process writing data into a table and another one that deduplicates rows every hour and rewrites a partition. If I push a bug to the dedup workflow, I will want to roll it back. Then I'd need to apply any appends that happened after the bad dedup run.